terrarium/terrarium.py

```diff
@@ -26,133 +24,55 @@
-        self._sensitive_arguments = set()
-
-    def _get_kwargs(self):
-        '''
-        Exclude arguments which are defined as being sensitive
-        '''
-        kwargs = super(Namespace, self)._get_kwargs()
-        return [
-            (name, value) for name, value in kwargs
-            if name not in self.sensitive_arguments
-        ]
-
-    @property
-    def sensitive_arguments(self):
-        return self._sensitive_arguments
-
-    def add_sensitive_arguments(self, *args):
-        '''
-        Define arguments as being "sensitive." That is, when this object is
-        converted to text, these attributes are not included.
-
-        `args` is a list of strings that indicate the attribute name.
-        '''
-        self._sensitive_arguments |= set(args)
-
-
-argparse.Namespace = Namespace
-
-
-def rmtree(path):
-    try:
-        if os.path.islink(path):
-            os.unlink(path)
-        elif os.path.isdir(path):
-            shutil.rmtree(path)
-        else:
-            os.unlink(path)
-        return True
-    except OSError, why:
-        logger.warn(
-            'Failed to remove %s. '
-            'Make sure you have permissions to this path. '
-            '%s' % (path, why)
-        )
-        return False
 
+PYTHONWARNINGS_IGNORE_PIP_PYTHON2_DEPRECATION = (
+    'ignore:DEPRECATION::pip._internal.cli.base_command'
+)
 
-# Helper method to determine the actual type of the file without relying on the
-# file extension
-def get_type(path):
-    with open(path) as f:
-        for file_type, magic in MAGIC_NUM.items():
-            f.seek(magic[1])
-            if magic[0] == f.read(len(magic[0])):
-                return file_type
-    return None
+PYTHONWARNINGS = [
+    PYTHONWARNINGS_IGNORE_PIP_PYTHON2_DEPRECATION,
+]
 
 
 class Terrarium(object):
-
     def __init__(self, args):
         self.args = args
         self._requirements = None
-        self._digest = None
 
-    @property
-    def digest(self):
-        if self._digest is not None:
-            return self._digest
-        m = hashlib.new(self.args.digest_type)
-        m.update('\n'.join(self.requirements))
-        self._digest = m.hexdigest()
-        return self._digest
+    def get_digest(self):
+        return calculate_digest_for_requirements(
+            digest_type=self.args.digest_type,
+            requirements=self.requirements,
+        )
 
     @property
     def requirements(self):
         if self._requirements is not None:
             return self._requirements
         lines = []
-        for arg in self.args.reqs:
-            if os.path.exists(arg):
-                with open(arg, 'r') as f:
-                    for line in f.readlines():
-                        line = line.strip()
-                        if line and not line.startswith('#'):
-                            lines.append(line)
+        for path in self.args.reqs:
+            if os.path.exists(path):
+                lines.extend(parse_requirements(path=path))
```

Comment:
TIL

Reply:
What did you learn?

Reply:
`list.extend(generator)`
this would "generate" the generator and dump it in a list.